Supervisor login

BenjaminGomori opened the pull request:
Created login and signup forms.
Users can sign up as regular users or supervisors
The difference is in the Boolean value stored in the is_supervisor column in the database.

Users sign-up and login using 'username' and 'password'. 
A user can not sign-up using a 'username' already in use.
All users' passwords are hashed using the 'bcrypt' library.

Only logged-in users/supervisors can submit an assessment.
Only logged-in supervisors can view the assessment list.
Sessions are used, using the 'redis' library, in the front-end routers to block these unauthorized actions.

To determine the react <NavLink> to be displayed, I make use of the local storage. If a user goes to the URL and types the URL of a page he is not authorized to send a request for - the page will display. In the case of the assessment form, the page displays regularly (but the form can not be submitted beyond the front-end router due to sessions), for the assessment list no data is displayed as the front-end router does send the request.

Success and failure  messages are displayed via an 'alert()' pop-up and guide the user with what the result of his actions was and what to do next. 

Login:
![image](https://user-images.githubusercontent.com/54749949/113383332-b435f080-9351-11eb-95a1-a9ef8624044e.png)


User View of Application:
![image](https://user-images.githubusercontent.com/54749949/113383381-ce6fce80-9351-11eb-9aee-6fcc4e0b032b.png)

Supervisor View of Application:
![image](https://user-images.githubusercontent.com/54749949/113383406-dc255400-9351-11eb-94cd-80e3798093fa.png)

Failure  message example:
![image](https://user-images.githubusercontent.com/54749949/113383582-3e7e5480-9352-11eb-9d52-122d561bba7d.png)

BenjaminGomori committed:
OCAT - Removed config.example.json

BenjaminGomori committed:
API -  added Users as a bookshelf model to the to microservices.database exports.

BenjaminGomori committed:
API - Added logic the routers\Users

BenjaminGomori committed:
APP Component - Added login route to <BrowserRouter>

BenjaminGomori committed:
Login Form - Added LoginForm component.

BenjaminGomori committed:
OCAT - Added LoginService.

BenjaminGomori committed:
OCAT - libs - Added logic to UserService.

BenjaminGomori committed:
OCAT - routes - Added Login route.

BenjaminGomori committed:
OCAT - utils - Added LoginRoute to index file exports.

BenjaminGomori committed:
OCAT - added rendering logic to LoginRouter for displaying the login page.

BenjaminGomori committed:
OCAT - Added the 'html'/ejs to the login view.

BenjaminGomori committed:
api - work in progress - microservice for idendenty-managment-service

BenjaminGomori committed:
API - Created logic for password.

BenjaminGomori committed:
API - routes/User -  sending user data back to front-end.

BenjaminGomori committed:
OCAT - need to make use of this... -Adding session middle wear.

BenjaminGomori committed:
OCAT- Navbar/Bullitin - Added state and logic to show/not different <NavLink>.

BenjaminGomori committed:
OCAT - work in progress - reloading page to make local storage affect state, many comments, for debugging.

BenjaminGomori committed:
OCAT - router - Added logic to continue send back the data, and many commented out console.log.

BenjaminGomori committed:
Added logic to store in local storage.

BenjaminGomori committed:
Dependencies added - express session, cookie-parser, and bcrypt.

BenjaminGomori committed:
Mostly small changes, such as syntax and function name.

BenjaminGomori committed:
index.js - Added 'redis' to application to manage sessions.

BenjaminGomori committed:
<DashboardBulletin> - Adjusted state managment in, and a logout function.

BenjaminGomori committed:
login.service - added functions to provide components with user state currently stored in local storage, and for logging out.

BenjaminGomori committed:
Front-end - Added LogoutRouter

BenjaminGomori committed:
Front-end,  LoginRouter - Added some session logic.

BenjaminGomori committed:
Removed redirection from <LoginForm> component.

BenjaminGomori committed:
app.bundle

BenjaminGomori committed:
small sytnax fixes.

BenjaminGomori committed:
small change

BenjaminGomori committed:
Testing to see that I block in the front-end Router -  the call for Assessments Lists for non supervisor users.

BenjaminGomori committed:
small fix from false to true.

BenjaminGomori committed:
Going back to the original code after testing.

BenjaminGomori commented on API/src/microservices/Assessment-Service/index.js:
```diff
@@ -34,7 +34,7 @@ exports.retrieve = () => {
       let assesmentList = await new Assessments().where({deleted_at: null}).fetchAll().catch(function (e) {
         console.log('error in retrieving assessments')
-        resolve(assesmentList.toJSON());
+        resolve();
```
Can this be empty?

BenjaminGomori commented on OCAT/client/resources/js/components/shared/services/login.service.js:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    static getCurrentUser() {
```
Providing local storage to the front-end components, was I able to provide data from the session instead?

wilsonpg commented on API/src/microservices/Identity-Management-Service/index.js:
```diff
@@ -0,0 +1,41 @@
+  return new Promise(async (resolve, reject) => { //eslint-disable-line
+    try {
+
+      let storedSupervisor = await new Users().where({id: 1}).fetch().catch(function (e) {
+        console.log('error in retrieving supervisor information')
+        resolve();
+      });
```
Can you explain what's going on here?  Why are you supplying a static id of 1 here?  I guess you're saying the first user is ALWAYS the supervisor?

BenjaminGomori replied:
Yes, the supervisor data was entered into the database directly, and is always the first record. 
I do check to see if  both the username and encrypted password match the supervisor login data sent by the form.

I only included the login form with out the register/sign up because that is how I understood the instructions.
I also added a supervisor user into the database table directly, as that is how I understood the instructions. 

Regarding the login for users, I agree, that currently, its not very logical to have it.

I will fix this, I am thinking of:

Allowing all users to sign up.
I will have 2 buttons in the sign up form. 1) User Sign Up , 2) Supervisor Sign Up 
The difference will be represented in the database in the 'is_supervisor' column as true or false.
Then when the user logs in, assuming his data is correct, he will have a different view depending on whether or not he is supervisor or not.

I will let you know once I completed these fixes.

wilsonpg reviewed (approved):
This looks okay.  There's some spelling issues and I'm a little confused with the process.  If I'm understanding correctly, then you are just checking if the person trying to login has the same username as the supervisor (id 1 in your users table).

In the past, folks have done things like allow users to register as supervisors, and then supervisors have a different view (i.e. access to the assessment list).  Or simply make every registered user a supervisor, and don't require login for regular users.  Or you can keep it how it is.  If it works the way you've written it, that's fine too.  Just explain in your presentation your mindset when developing this.

BenjaminGomori committed:
Small chanegs.

BenjaminGomori committed:
Bookshelf, Users table - Allowing fetch  to retuen 0 records. This is used to make sure username is not already used when creating new user.

BenjaminGomori committed:
Adjusting microservices logic to allow dyncamic signUp.

BenjaminGomori committed:
API , routes - Adjusting the  ResponseHandler data coming back.

BenjaminGomori committed:
API, Password - Added logic to provide new hashed passwords to enable new user sign up.
Was not able to figure out how to return the hashed password without using a callback function.

BenjaminGomori committed:
<DashboardBulletin> - Corrected typo.

BenjaminGomori committed:
<LoginForm> - Added sign up form capabilities.

BenjaminGomori committed:
FE - AssessmentService - Added logic  to only allow logged in users to submit assessment.

BenjaminGomori committed:
FE - LoginService - Added sign up logic.
And added ogic to handle bad logging in.

BenjaminGomori committed:
OCAT - libs, UserService -  added sign up logic.

BenjaminGomori committed:
OCAT- api/assessment/router - added logic to disable assessment form submission from not logged in users.

BenjaminGomori committed:
OCAT , Login route -  added sin up logic.
 And changed the session logic for signing in.

BenjaminGomori committed:
Removed console logs and empty lines beofre code review.

BenjaminGomori committed:
Download Code Spell Checker, and corrected spelling errors

BenjaminGomori commented on API/src/utils/Password/index.js:
```diff
@@ -0,0 +1,32 @@
+    //   console.log('result');
+    //   console.log(result);
+    //   resolve(''+result);
+    //   return result;
```
I was not able to make this return statement return the actual Boolean value, that is why I am using a call back function.

wilsonpg reviewed (approved):
This looks great!  approved
